File: static/solutions/cpp/student_sort.cc
// Copyright (c) 2015 Elements of Programming Interviews. All rights reserved.

#include <algorithm>
#include <cassert>
#include <string>
#include <vector>

using std::sort;
using std::string;
using std::vector;

// @include
struct Student {
  bool operator<(const Student& that) const { return name < that.name; }

  string name;
  double grade_point_average;
};

void SortByGPA(vector<Student>* students) {
  sort(students->begin(), students->end(),
       [](const Student& a, const Student& b) {
         return a.grade_point_average >= b.grade_point_average;
       });
}

void SortByName(vector<Student>* students) {
  // Rely on the operator< defined in Student.
  sort(students->begin(), students->end());
}
// @exclude

int main(int argc, char* argv[]) {
  vector<Student> students = {{"A", 4.0}, {"C", 3.0}, {"B", 2.0}, {"D", 3.2}};
  SortByName(&students);
  assert(is_sorted(students.begin(), students.end()));
  SortByGPA(&students);
  assert(is_sorted(students.begin(), students.end(),
                   [](const Student& a, const Student& b) {
                     return a.grade_point_average >= b.grade_point_average;
                   }));
  return 0;
}

File: static/cpp/java/src/main/java/com/epi/StudentSort.java
// Copyright (c) 2015 Elements of Programming Interviews. All rights reserved.

package com.epi;

import java.util.Arrays;
import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;
import java.util.List;

public class StudentSort {
  // @include
  public static class Student implements Comparable<Student> {
    public String name;
    public double gradePointAverage;

    public int compareTo(Student that) { return name.compareTo(that.name); }

    Student(String name, double gradePointAverage) {
      this.name = name;
      this.gradePointAverage = gradePointAverage;
    }
  }

  public static void sortByGPA(List<Student> students) {
    Collections.sort(
        students, Collections.reverseOrder(new Comparator<Student>() {
          @Override
          public int compare(Student a, Student b) {
            return Double.compare(a.gradePointAverage, b.gradePointAverage);
          }
        }));
  }

  public static void sortByName(List<Student> students) {
    Collections.sort(students);
  }
  // @exclude

  public static void main(String[] args) {
    List<Student> students
        = Arrays.asList(new Student("A", 4.0), new Student("C", 3.0),
                        new Student("B", 2.0), new Student("D", 3.2));
    sortByName(students);
    for (int i = 1; i < students.size(); ++i) {
      assert(students.get(i - 1).name.compareTo(students.get(i).name) <= 0);
    }
    sortByGPA(students);
    for (int i = 1; i < students.size(); ++i) {
      assert(students.get(i - 1).gradePointAverage
             >= students.get(i).gradePointAverage);
    }
  }
}
